include/xgboost/tree_model.h

```diff
@@ -724,38 +723,5 @@ inline bool RegTree::FVec::IsMissing(size_t i) const {
 inline bool RegTree::FVec::HasMissing() const {
   return has_missing_;
 }
-
-template <bool has_missing>
-inline int RegTree::GetLeafIndex(const RegTree::FVec& feat) const {
-  bst_node_t nid = 0;
-  while (!(*this)[nid].IsLeaf()) {
-    unsigned split_index = (*this)[nid].SplitIndex();
-    nid = this->GetNext<has_missing>(nid, feat.GetFvalue(split_index),
-                                     has_missing && feat.IsMissing(split_index));
-  }
-  return nid;
-}
-
-/*! \brief get next position of the tree given current pid */
-template <bool has_missing>
-inline int RegTree::GetNext(int pid, bst_float fvalue, bool is_unknown) const {
-  if (has_missing) {
-    if (is_unknown) {
-      return (*this)[pid].DefaultChild();
-    } else {
-      if (fvalue < (*this)[pid].SplitCond()) {
-        return (*this)[pid].LeftChild();
-      } else {
-        return (*this)[pid].RightChild();
-      }
-    }
-  } else {
-    // 35% speed up due to reduced miss branch predictions
-    // The following expression returns the left child if (fvalue < split_cond);
-    // the right child otherwise.
-    return (*this)[pid].LeftChild() + !(fvalue < (*this)[pid].SplitCond());
-  }
```

Comment:
Why it's not preserved in new `GetNextNode` function?
At least 35% of performance regression will be introduced
Fast check on higgs1m data: **master 0.77s vs this pr: 1.24s** (caused by a lot of newly introduced conditions).

Reply:
if it's possible I could help with reducing the perf degradations :)

Reply:
Sure, feel free to push to my branch or made PRs. ;-)

I did run some quick benchmarks on this PR with datasets generated by `sklearn.datasets.make_regression` with 1e6 samples and 30 features.  The run time didn't change at all ...

Reply:
@ShvetsKS Also please be mindful that we have a pruner, I'm not entirely sure whether right node index is exactly left node index + 1.

Reply:
@ShvetsKS I came up with this branch, feel free to take a look https://github.com/trivialfis/xgboost/tree/cat-cpu-predictor-optimization .

I'm running it on `make_regression` with 1e7 samples and 30 features, the run time is about the same (slightly faster).  Not sure whether it works on you environment.

Reply:
possibly you can't observe regression as dmatrix create takes most of the time for inference on such dimensions, for clear comparison it's better to use `inplace_predict`

Reply:
> @ShvetsKS I came up with this branch, feel free to take a look https://github.com/trivialfis/xgboost/tree/cat-cpu-predictor-optimization .

I'll try to be helpful :)

Reply:
Actually I'm using `inplace_predict`.  This is the function is used to run the simple bench (it's part of some other random scripts I haven't been able to cleanup yet ...)

``` python
def reg_predict(args):
    X, y = make_regression(args)
    parameters = configure(args, PARAMETERS.copy())
    booster = xgb.Booster(
        model_file=parameters.get('tree_method', 'updater') + "reg.json"
    )
    s = time()
    for i in range(10):
        booster.inplace_predict(X)
    print("inplace_predict", time() - s)
```

Reply:
@ShvetsKS I have merged my small optimization into this branch.

Reply:
@trivialfis  sorry for long response!
I fixed perf degradations with simple modifications: https://github.com/ShvetsKS/xgboost/commits/cat-cpu-predictor 907ce2e7d98cc1b1a59a41c54a28d90af17416d6

1) moved template parameter setting for `has_categorical` from the loop
2) usage of formula `left node index + 1` was reverted as we have the check for this (https://github.com/dmlc/xgboost/blob/master/src/tree/updater_quantile_hist.cc#L1067) and pruners do not change the node id in case of pruning. Seems there is no reason not to use this formula as in pruning only `SetLeaf` is called and it doesn't change the id on node.
3) small improvement of `GetNextNode` as there is no need to use all nodes as only one node reference is needed

results:
| Data        | master | this pr | fixed perf |
| ------------- |:-------------:| -----:| -----:|
| higgs     | 0.616 | 1.03 | 0.612 |
| airline-ohe      | 0.78 |  0.67 | 0.79 |
| mortgage |2.41|2.71|2.41|
| msrank |1.90|2.57|1.89|
| santander|0.272|0.41|0.23|

Reply:
> usage of formula left node index + 1 was reverted

Users can run pruner themselves, it doesn't have to be part of hist.

Reply:
The issue with pruner is, those deleted nodes can be reused in `AllocNode`.

Reply:
Rest of your optimization looks good and feel free to push to my branch.  ;-)  Thanks for working on this

Reply:
> > usage of formula left node index + 1 was reverted
> 
> Users can run pruner themselves, it doesn't have to be part of hist.

strange there is no any compliance since this approach was introduced long ago.
Could you provide more details or some example with such situations, and I think it's possible to add checks into `AllocNode` for example?

Reply:
What's the overhead of calling right node method given a `Node` struct (instead of indexing into tree) first?  It should be fully inlined right?

Reply:
sorry, on higgs with `inplace_predict` 1.5x degradation: use `RightChild()` with exact condition - 0.927s, `LeftChild() + 1`without exact condition - 0.615s.
seems the reason is branch prediction miss, not inlining.

Reply:
Thanks for sharing the perf change.  So far it seems to be correct since after a tree is pruned, there's no way to continue growing it so the `Reuse` function is never used.  As far as I know, non of the existing updater can do it at the moment.

Reply:
@ShvetsKS I have merged your code into my branch and removed the unused variable `go_left`.  Could you please take another look?

Reply:
@trivialfis thanks a lot. Looks good for me!

Reply:
Thanks for sharing the optimization!